bundler/lib/bundler/gem_helpers.rb

```diff
@@ -55,7 +56,8 @@ def select_best_platform_match(specs, platform)
     class PlatformMatch
       def self.specificity_score(spec_platform, user_platform)
         return -1 if spec_platform == user_platform
-        return 1_000_000 if spec_platform.nil? || spec_platform == Gem::Platform::RUBY || user_platform == Gem::Platform::RUBY
+        return 1_000_000 if spec_platform.nil? || spec_platform == Gem::Platform::RUBY || user_platform == Gem::Platform::RUBY # <= last one is weird, not sure it should happen
```

Comment:
Here I'm not seeing when `user_platform == Gem::Platform::RUBY` is a legit case:

- for `ruby` gem and runtime platform, it would be covered by `spec_platform == user_platform` 
- it would just let through any non-`ruby` gem on `ruby` runtime platform†
- it would just let through any non-`ruby` gem on any runtime platform, given `spec_group_ruby` adds a `ruby` runtime platform processing step

So I thought, `1_000_000` is ultra low priority (IIUC), but maybe `return nil` when it's a definitive not-a-match-at-all, then `.compact!` the `sort_by` result would give only those `sorted_matching` that truly have a chance to match, without the ones we are sure we do not want.

† assuming there is any before `specificity_score` is called, which seems like it should often be the case, except a second question seems to ram some into this.

Reply:
I _think_ it can be assumed that it's only for sorting but shouldn't remove any specs. Filtering out specs that should never match should be done previously to using this method.

Just to try to give you some context from what I recall from fighting against these issues, this method is used in a few different situations, where "user_platform" doesn't always mean "runtime platform" but a platform we are resolving dependencies or materializing a resolution against. Your reasoning might still make sense, but just wanted to clarify that. For example:

* We might be on linux, but resolve against java. This is what happens when you do `bundle lock --add-platform java`, or when you have a lockfile including JAVA in the list of platforms. In those cases, a resolution is found that not only matches the running platform, but also JAVA.

* When I say materializing I mean picking all the specs in a resolution (if recorded, in the lockfile) and selecting only those which are relevant for a specific platform. For example, the lockfile can include specs for different platforms at the same time, like `foo-1.0` and `foo-1.0-x86_64-linux` if the lockfile includes both say `x86_64-linux` and `ruby` in the `PLATFORMS` sections. It can even include specs that are only relevant for some platforms, for example, if `foo-1.0-x86_64-linux` depends on `bar`, but `foo-1.0` does not depend on `bar`.

Also, and still not sure whether it's relevant to these issues, but just to give some context on this "ruby spec_group". Since different platform specific variants can have different dependencies, the resolver always adds this fallback spec group, so that it can consider all possibilities. So in the example I gave above, when resolving for `x86_64-linux`, one resolution possibility would be the `<foo-1.0, x86_64_linux>` spec group, an another one would be the `<foo-1.0, ruby>` fallback spec group, and they need to be considered as different possibilities because one of them brings in the `bar` dependency requirement while the other one does not.

Not sure if all of the above answers the question but hopefully brings some light.

Reply:
Yes it does! It was especially helpful as I had to dig in deep to [produce a patch](https://github.com/rubygems/rubygems/issues/3174#issuecomment-963166985) specifically for bundler 1.17.